### PR TITLE
Phase 5: Born-Again Networks — Self-Distillation for Better Generalization (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Train the baseline model normally (teacher), then train a second identical model (student) using a weighted combination of the ground truth loss AND the teacher's predictions as soft targets. The student learns to replicate the teacher's "dark knowledge" — smoothed prediction surfaces that regularize training and improve generalization.

**Why this should work:**
- Born-Again Networks (Furlanello et al., ICML 2018) showed that self-distillation consistently improves accuracy even with identical architectures. The student outperforms the teacher because soft targets provide a richer gradient signal.
- Our p_in has high variance (std=0.58 across 8 seeds), suggesting the model converges to different local minima. Self-distillation provides a smoother loss landscape via the teacher's averaged predictions.
- Particularly useful for OOD generalization (p_oodc, p_re) — the teacher's predictions capture learned physics relationships that raw labels don't.
- We already have EMA as a form of temporal ensembling. Self-distillation is the spatial complement — averaging predictions across the training distribution.

**References:**
- Furlanello et al., "Born Again Neural Networks" (ICML 2018)
- Hinton et al., "Distilling the Knowledge in a Neural Network" (2015)
- Zhang et al., "Self-Distillation as Instance-Specific Label Smoothing" (NeurIPS 2020)

## Instructions

See first comment for implementation details.

## Baseline (8-seed)
val/loss 0.404+/-0.004, p_in 13.33+/-0.58, p_oodc 8.37+/-0.22, p_tan 33.57+/-0.44, p_re 24.58+/-0.13

### CODE REVIEW

**Comment by tcapelle** [comment]
## Detailed Implementation

### Two-Phase Training

**Phase 1 (Teacher):** Normal baseline training for 90 min (half budget). Save EMA checkpoint.
**Phase 2 (Student):** Load fresh model, train for remaining 90 min with distillation loss.

### 1. Add config flags

```python
self_distill: bool = False         # enable self-distillation
distill_alpha: float = 0.5         # weight of distillation loss (1-alpha = ground truth weight)
distill_teacher_path: str = ""     # path to teacher checkpoint (auto-generated if empty)
```

### 2. Implementation

```python
if cfg.self_distill:
    # Phase 1: Train teacher for first half of budget
    # (Use existing training loop, save checkpoint at halfway)
    
    if not cfg.distill_teacher_path:
        # Train normally for half the epochs/time, save teacher
        # After half the budget, save model:
        teacher_path = f"/tmp/teacher_{cfg.seed}.pt"
        torch.save(ema_model.state_dict(), teacher_path)
        
        # Phase 2: Reset student, load teacher for distillation
        teacher = copy.deepcopy(model)
        teacher.load_state_dict(torch.load(teacher_path))
        teacher.eval()
        for p in teacher.parameters():
            p.requires_grad = False
        
        # Reset student model
        model = Transolver(cfg).to(device)
        optimizer = create_optimizer(model, cfg)
        scheduler = create_scheduler(optimizer, cfg)
```

### 3. Distillation loss

```python
if cfg.self_distill and teacher is not None:
    with torch.no_grad():
        teacher_pred = teacher(batch)["preds"]
    
    # Ground truth loss (existing)
    gt_loss = compute_surface_loss(pred, y_norm, surf_mask)
    
    # Distillation loss: MSE between student and teacher predictions on surface
    distill_err = ((pred - teacher_pred) ** 2 * surf_mask.unsqueeze(-1)).sum() / surf_mask.sum().clamp(min=1)
    
    # Combined loss
    loss = (1 - cfg.distill_alpha) * gt_loss + cfg.distill_alpha * distill_err
```

### Alternative: Online Self-Distillation

Instead of two phases, use the EMA model as a continuously updated teacher:

```python
if cfg.self_distill_online:
    with torch.no_grad():
        teacher_pred = ema_model(batch)["preds"]
    
    distill_loss = ((pred - teacher_pred.detach()) ** 2 * surf_mask.unsqueeze(-1)).sum() / surf_mask.sum().clamp(min=1)
    loss = loss + cfg.distill_alpha * distill_loss
```

This is simpler (no two-phase training) and the EMA already provides a smoothed teacher.

### GPU Sweep (8 parallel)

| GPU | Config | wandb_name |
|-----|--------|------------|
| 0 | Online distill from EMA, alpha=0.1 | fern/distill-ema-a01 |
| 1 | Online distill from EMA, alpha=0.3 | fern/distill-ema-a03 |
| 2 | Online distill from EMA, alpha=0.5 | fern/distill-ema-a05 |
| 3 | Two-phase, alpha=0.3 | fern/distill-2phase-a03 |
| 4 | Two-phase, alpha=0.5 | fern/distill-2phase-a05 |
| 5 | Two-phase, alpha=0.7 (teacher-heavy) | fern/distill-2phase-a07 |
| 6 | Baseline seed 42 | fern/baseline-s42 |
| 7 | Baseline seed 43 | fern/baseline-s43 |

All use `--wandb_group phase5/self-distillation`.

Base command:
```bash
python train.py --agent fern --wandb_name "fern/<variant>" \
  --wandb_group phase5/self-distillation \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep
```

### Notes
- The online EMA distillation (GPUs 0-2) is simpler and doesn't waste epochs on teacher training
- For two-phase, the teacher should train for ~80 epochs (half budget), then student for ~80 epochs
- Use MSE for distillation loss (smoother gradients than L1 for soft targets)
- The distillation loss should be on SURFACE predictions only (consistent with our metric)

**Comment by tcapelle** [comment]
## Results: Self-Distillation Sweep (8 GPUs, 3 hours each)

All 8 runs completed. W&B group: `phase5/self-distillation`

### Summary Table

| GPU | Config | val/loss | p_in | p_oodc | p_tan | p_re | Best Epoch |
|-----|--------|----------|------|--------|-------|------|------------|
| 0 | Online EMA, α=0.1 | 0.4094 | 13.2 | 8.6 | 34.2 | 24.8 | 153 |
| 1 | Online EMA, α=0.3 | 0.4026 | 12.9 | 8.6 | 33.5 | 24.5 | 153 |
| 2 | Online EMA, α=0.5 | 0.4025 | 13.1 | 8.4 | 33.8 | 24.6 | 153 |
| 3 | Two-phase, α=0.3 | 0.5063 ❌ | 21.3 | 13.9 | 36.6 | 28.1 | 80 |
| 4 | Two-phase, α=0.5 | 0.5104 ❌ | 21.7 | 12.5 | 36.4 | 27.9 | 80 |
| 5 | Two-phase, α=0.7 | 0.5206 ❌ | 21.8 | 14.0 | 38.3 | 28.0 | 79 |
| 6 | **Baseline s42** | **0.4001** | 12.6 | 8.3 | 33.1 | 24.5 | 160 |
| 7 | **Baseline s43** | **0.3992** ⭐ | 13.0 | 8.2 | 33.4 | 24.5 | 160 |

### Analysis

**Baseline wins** again. Mean baseline val/loss = 0.3997.

**Online EMA distillation (GPUs 0-2):**
- α=0.3 and α=0.5 are nearly identical (0.4026, 0.4025) — marginally worse than baseline (0.3997)
- α=0.1 is the worst online config (0.4094) — too little distillation signal to matter, but enough to disrupt training
- The online distillation doesn't help because the EMA model and the training model are too similar — the "dark knowledge" is nearly identical to what the student already learns

**Two-phase distillation (GPUs 3-5) — catastrophic failure:**
- All three configs are dramatically worse (val/loss 0.50-0.52 vs 0.40 baseline)
- Best model selected at epoch 79-80 — this is the teacher's checkpoint from Phase 1 (before reset)
- The student model (Phase 2) never catches up — it starts from random init at the halfway point and only gets ~60 epochs to train
- The torch.compile recompilation after model reset eats ~50s per epoch, further reducing available training time
- With only ~60 epochs and distillation pulling toward a weak teacher, the student can't converge

**Why self-distillation fails here:**
1. **EMA is too similar**: Unlike classification where teacher/student representations differ meaningfully, our EMA teacher's predictions are almost identical to the student's within a few epochs. The distillation signal is redundant.
2. **Half-budget teacher is weak**: The teacher at epoch 80 (val/loss ~0.50) hasn't converged enough to provide useful soft targets. Born-again networks assume a fully-trained teacher.
3. **Recompilation cost**: The model reset + torch.compile warmup wastes ~50s, equivalent to ~4 epochs of lost training.

**Verdict:** Self-distillation does not improve over baseline. Close this PR.

### W&B Links
- Online: [α=0.1](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/wnopgnem) [α=0.3](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mcfe39ux) [α=0.5](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/c707wt3t)
- Two-phase: [α=0.3](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/wpkei9tu) [α=0.5](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3zickm6c) [α=0.7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/zrgj5w34)
- Baselines: [s42](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/7i4b526s) ⭐ [s43](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mocuy7k8)
